### PR TITLE
Switch from handling output by hand to using a custom registry

### PR DESCRIPTION
**Pull request by Conorbro**
Completes #76

### CODE REVIEW

File: dns.go [hunk 1]
```diff
@@ -81,15 +81,36 @@ func validRcode(rcode int, valid []string) bool {
...
+	})
+	probeDNSAuthorityRRSGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "probe_dns_authority_rrs",
+		Help: "Returns number of entries in the authority resource record list",
+	})
+	probeDNSAdditionalRRSGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "probe_dns_additional_rrs",
+		Help: "Returns number of entries in the additional resource record list",
+	})
+	registry.Register(probeIPProtocolGauge)
```
Comment (brian-brazil) [inline dns.go]: This should use MustRegister, so we'll catch future bugs.

File: dns_test.go [hunk 2]
```diff
@@ -249,14 +259,30 @@ func TestAuthoritativeDNSResponse(t *testing.T) {
...
+			}
+			var buf bytes.Buffer
+			for _, mf := range mfs {
+				if _, err := expfmt.MetricFamilyToText(&buf, mf); err != nil {
+					t.Fatal(err)
+				}
+			}
+
+			for _, re := range []*regexp.Regexp{
+				regexp.MustCompile("\nprobe_dns_answer_rrs 1"),
```
Comment (brian-brazil) [inline dns_test.go]: Why the leading \n?

File: http.go [hunk 3]
```diff
@@ -56,10 +56,54 @@ func matchRegularExpressions(reader io.Reader, config HTTPProbe) bool {
...
+		})
+
+		redirectsGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "probe_http_redirects",
+			Help: "The number of redirects",
+		})
+
+		isSSLGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "probe_http_ssl",
+			Help: "Checks if SSL cert is valid",
```
Comment (brian-brazil) [inline http.go]: This is if SSL was used

File: http.go [hunk 4]
```diff
@@ -56,10 +56,54 @@ func matchRegularExpressions(reader io.Reader, config HTTPProbe) bool {
...
+		})
+
+		probeIPProtocolGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "probe_ip_protocol",
+			Help: "Specifies whether probe ip protocl is IP4 or IP6",
+		})
+
+		probeSSLEarliestCertExpiryGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "probe_ssl_earliest_cert_expiry",
+			Help: "Returns earliest SSL cert expiry date",
```
Comment (brian-brazil) [inline http.go]: This is a unixtime, not a date

File: http.go [hunk 5]
```diff
@@ -56,10 +56,54 @@ func matchRegularExpressions(reader io.Reader, config HTTPProbe) bool {
...
+			Name: "probe_http_ssl",
+			Help: "Checks if SSL cert is valid",
+		})
+
+		statusCodeGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "status_code",
+			Help: "Response HTTP status code",
+		})
+
+		sslExpiry = prometheus.NewGauge(prometheus.GaugeOpts{
```
Comment (brian-brazil) [inline http.go]: This is unused

File: http.go [hunk 6]
```diff
@@ -56,10 +56,54 @@ func matchRegularExpressions(reader io.Reader, config HTTPProbe) bool {
...
+			Help: "The number of redirects",
+		})
+
+		isSSLGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "probe_http_ssl",
+			Help: "Checks if SSL cert is valid",
+		})
+
+		statusCodeGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "status_code",
```
Comment (brian-brazil) [inline http.go]: probe_http_

File: http.go [hunk 7]
```diff
@@ -188,17 +233,18 @@ func probeHTTP(target string, w http.ResponseWriter, module Module) (success boo
...
 	}
-	fmt.Fprintf(w, "probe_http_status_code %d\n", resp.StatusCode)
-	fmt.Fprintf(w, "probe_http_content_length %d\n", resp.ContentLength)
-	fmt.Fprintf(w, "probe_http_redirects %d\n", redirects)
-	fmt.Fprintf(w, "probe_http_ssl %d\n", isSSL)
+
+	statusCodeGauge.Set(float64(resp.StatusCode))
+	contentLengthGauge.Set(float64(resp.ContentLength))
+	redirectsGauge.Set(float64(redirects))
+	isSSLGauge.Set(float64(isSSL))
```
Comment (brian-brazil) [inline http.go]: This can go above now, eliminating the isSSL variable.

File: main.go [hunk 8]
```diff
@@ -61,8 +63,8 @@ type HTTPProbe struct {
 	FailIfMatchesRegexp    []string          `yaml:"fail_if_matches_regexp"`
 	FailIfNotMatchesRegexp []string          `yaml:"fail_if_not_matches_regexp"`
 	TLSConfig              config.TLSConfig  `yaml:"tls_config"`
-	Protocol               string            `yaml:"protocol"`              // Defaults to "tcp".
-	PreferredIPProtocol    string            `yaml:"preferred_ip_protocol"` // Defaults to "ip6".
+	Protocol               string            `yaml:"protocol"`              // Defaults to TCP.
```
Comment (brian-brazil) [inline main.go]: This defaults to "tcp", not TCP

File: main.go [hunk 9]
```diff
@@ -153,20 +163,26 @@ func probeHandler(w http.ResponseWriter, r *http.Request, conf *Config) {
...
+	registry.MustRegister(probeSuccessGauge)
+	registry.MustRegister(probeDurationGauge)
+	success := prober(target, w, module, registry)
+	probeDurationGauge.Set(time.Since(start).Seconds())
 	if success {
-		fmt.Fprintln(w, "probe_success 1")
+		probeSuccessGauge.Set(1)
 	} else {
-		fmt.Fprintln(w, "probe_success 0")
+		probeSuccessGauge.Set(0)
```
Comment (brian-brazil) [inline main.go]: It's initialised to 0, so you don't need the else

File: http.go [hunk 10]
```diff
@@ -56,10 +56,48 @@ func matchRegularExpressions(reader io.Reader, config HTTPProbe) bool {
...
+		})
+
+		redirectsGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "probe_http_redirects",
+			Help: "The number of redirects",
+		})
+
+		isSSLGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "probe_http_ssl",
+			Help: "Checks if SSL was used",
```
Comment (brian-brazil) [inline http.go]: Indicates, not checks. It's also for the final redirect.

**Comment by grobie** [comment]
Have you measured the impact of the overhead of creating all metrics on the fly during each probe? Why isn't it possible to just use const metrics here?

**Pull request closed without merging**